packages/api/endpoints/rules.js

```diff
@@ -67,29 +66,36 @@ async function get(req, res) {
  */
 async function post(req, res) {
   const {
-    model = new models.Rule(),
-    dbClient = await getKnexClient(),
+    ruleModel = new models.Rule(),
+    rulePgModel = new RulePgModel(),
+    knex = await getKnexClient(),
+    esClient = await Search.es(),
   } = req.testContext || {};
 
   let record;
   const apiRule = req.body || {};
   const name = apiRule.name;
-  const rulePgModel = new RulePgModel();
 
-  if (await model.exists(name)) {
+  if (await ruleModel.exists(name)) {
     return res.boom.conflict(`A record already exists for ${name}`);
   }
 
   try {
     apiRule.createdAt = Date.now();
     apiRule.updatedAt = Date.now();
-    const postgresRule = await translateApiRuleToPostgresRule(apiRule, dbClient);
-
-    await dbClient.transaction(async (trx) => {
-      await rulePgModel.create(trx, postgresRule);
-      record = await model.create(apiRule);
-    });
-    if (inTestMode()) await addToLocalES(record, indexRule);
+    const postgresRule = await translateApiRuleToPostgresRule(apiRule, knex);
+
+    try {
+      await knex.transaction(async (trx) => {
+        await rulePgModel.create(trx, postgresRule);
+        record = await ruleModel.create(apiRule);
+        await indexRule(esClient, record, process.env.ES_INDEX);
+      });
+    } catch (innerError) {
+      // Clean up DynamoDB record in case of any failure
+      await ruleModel.delete(apiRule);
```

Comment:
Is this right? If `await rulePgModel.create(trx, postgresRule);` fails the transaction will fail and throw. This would cause the Dynamo model to delete even if though we didn't try to create it I think.

Reply:
This would apply to all of the POST endpoints.

Reply:
>This would cause the Dynamo model to delete even if though we didn't try to create it I think.

It will certainly try to do a delete even though we didn't succeed in creating it, but since DELETE is idempotent, trying to delete something that doesn't exist shouldn't be an issue. It is perhaps wasteful, but to handle it more elegantly we'd have to inspect the type of error that we catch which adds a lot of complexity.

In general these extra `catch` blocks are kinda unfortunate, but it's a necessary evil with three operations in the transaction

Reply:
I see what you mean. I think you're right. If line 91 or 92 fails, we'll try to delete a Dynamo Rule that doesn't exist and it'll just return 0 rows. That doesn't seem so bad.